src/OrleansRuntime/Facet/GrainFacetPropertyInjector.cs

```diff
@@ -0,0 +1,70 @@
+﻿
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Orleans.Runtime
+{
+    internal class GrainFacetPropertyInjector
+    {
+        // TODO: consider storing in grain type data and constructing at startup to avoid runtime errors. - jbragg
+        /// <summary>
+        /// Cached property injector by type
+        /// </summary>
+        private readonly ConcurrentDictionary<Type, Injector> injectors;
+
+        public GrainFacetPropertyInjector()
+        {
+            injectors = new ConcurrentDictionary<Type, Injector>();
+        }
+        
+        public void InjectProperties(IGrainActivationContext grainActivationContext)
+        {
+            Injector injector = injectors.GetOrAdd(grainActivationContext.GrainType, type => new Injector(type));
+            injector.InjectProperties(grainActivationContext);
+        }
+
+        /// <summary>
+        /// Injects grain facets into properties of an object
+        /// </summary>
+        private class Injector
+        {
+            private readonly List<Action<IGrainActivationContext>> propertyInjectors;
+
+            public Injector(Type type)
+            {
+                this.propertyInjectors = CreatePropertyInjectors(type).ToList();
+            }
+
+            public void InjectProperties(IGrainActivationContext grainContext)
+            {
+                propertyInjectors.ForEach(injector => injector.Invoke(grainContext));
```

Comment:
Using `foreach` here would avoid these two per-activation allocations: one for the closure to capture `grainContext` and one for the conversion from method to `Func<Action<IGrainActivationContext>>`.